test/integration/dataplane_mtls_test.go

```diff
@@ -0,0 +1,118 @@
+//go:build dataplane_tests
+// +build dataplane_tests
+
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package integration
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+var (
+	CAFile         = certFile("ca.pem")
+	ClientCertFile = certFile("client.pem")
+	ClientKeyFile  = certFile("client-key.pem")
+	ServerCertFile = certFile("server.pem")
+	ServerKeyFile  = certFile("server-key.pem")
+)
+
+// TestGRPCClient tests the gRPC client against a running Docker container.
+func TestGRPCClient(t *testing.T) {
+	// Get the current working directory
+	dir, err := os.Getwd()
+	if err != nil {
+		fmt.Println("Error:", err)
+		return
+	}
+
+	fmt.Println("Current working directory:", dir)
+	clientTLSConfig, err := setupTLSConfig(TLSConfig{
+		CAFile:   CAFile,
+		CertFile: ClientCertFile,
+		KeyFile:  ClientKeyFile,
+	})
+	require.NoError(t, err)
+
+	clientCreds := credentials.NewTLS(clientTLSConfig)
+	conn, err := grpc.Dial("localhost:9874", grpc.WithTransportCredentials(clientCreds))
+	require.NoError(t, err)
+	defer conn.Close()
+}
+
+// Helper functions
+func certFile(filename string) string {
+	wd, _ := os.Getwd()
+	projectDir := filepath.Clean(filepath.Join(wd, "..", ".."))
+	if certsDir := os.Getenv("TEST_CERTS_PATH"); certsDir != "" {
+		return filepath.Join(projectDir, certsDir, filename)
+	}
+	panic("Env var TEST_CERTS_PATH not found. Please specify path to mTLS test certs")
+}
+
+func setupTLSConfig(cfg TLSConfig) (*tls.Config, error) {
```

Comment:
There's a bunch of optionality present in here, in that it will try and generate certificates if not present, e.t.c.. However (if I'm not misunderstanding) we actually currently expect for all runs of this test that all values are present? If so, and because this is a private function only used here for this purpose, we may actually just want to _fail_ when things are not present instead as this will help us more quickly realize when we've broken our build system. Wdyt?

Reply:
This function does not generate certs if they are not absent, it just reads in the certs and converts them into the tls.config struct that grpc expects.

However, it doesn't abort early if paths to CertFile, KeyFile or CAFile are not present. I can add those.